ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,258 @@
+(KeyHash 'DRep c)`.
+
+- The reward update
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesRu :: NewEpochState era -> StrictMaybe (PulsingRewUpdate (EraCrypto era)) )
+  
+data RewardUpdate c = RewardUpdate
+  { deltaT    :: !DeltaCoin
+  , deltaR    :: !DeltaCoin
+  , rs        :: !(Map (Credential 'Staking c) (Set (Reward c)))
+  , deltaF    :: !DeltaCoin
+  , nonMyopic :: !(NonMyopic c)
+  }
+```
+
+Where the `PulsingRewUpdate` is just a mechanism in order to pulse through the
+stake snapshot and in the end yield a `RewardUpdate`.
+
+## Dynamics of these maps
+
+### Unified map
+
+On each application of a block (i.e. the `BBODY` rule) and each tick (i.e. the
+`TICK` rules) we know what entries of the unified map are needed to execute the
+rule. Therefore it fits the current design of UTxO-HD, and therefore can follow
+the _same_ pattern as we currently have for the UTxO set:
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+  
+In particular, we know that the sets required for the deltas on the unified map
+are small and therefore fit the overall design.
+
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the unified map. Perhaps even 4 tables or a table of triplets. It
+> will have its own place on the `DbChangelog` too. The flow above describes the
+> general strategy for calling the `BBODY` rule, i.e. `(re)applyBlockOpts` and
+> for calling the `TICK` rule, i.e. `applyTickOpts`.
+
+### Stake distribution
+
+The update to this map is performed by `updateStakeDistribution`. It is known
+before calling the `BBODY` rule which UTxOs are going to be deleted and the rule
+execution logic itself knows the UTxOs that are going to be added. Therefore we
+can follow a logic similar to the above:
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the incremental stake distribution (maybe two tables, one for creds
+> one for ptrs). It will have its own place on the `DbChangelog` too. The flow
+> above describes the general strategy for calling the `BBODY` rule, i.e.
+> `(re)applyBlockOpts`.
```

Comment:
The incrementally maintained stake distribution should be a table where we normally only need to do monoidal updates. This means we turn a read/write workload into a write only workload. This is covered in the original report and is the motivation for the inclusion of monoidal update tables (including in the prototype).

This is important for performance, otherwise we are going to be doing almost as many lookups in the stake distribution table as we do lookups in the utxo table. Lookups are the bottleneck for a write-optimised disk backend.

Reply:
Are you suggesting that we do something more like double entry booking, where we (monoidally) send updates to two separate things, additions and deletions, and only combine them when needed?

Reply:
I'm not sure about two separate things (what two?) but yes I am suggesting a table that can support monoidal updates, as well as insert/delete/lookup.

So for an incrementally updated stake distribution, that maps stake address to lovelace, it would be updated by the ledger working out the delta in stake arising for a stake address from a tx, and that is the table update. So in particular, the lovelace value of a spent input is a negative delta for that input's stake address, while lovelace outputs contribute positively to the balance of their corresponding stake addresses. Note that no lookup is needed: the sum total of the lovelace for a stake address is not needed.

And the practical consequence is the update-only workload is a write-only workload, which is much cheaper than a read/write workload.